### PR TITLE
[Merged by Bors] - feat(GroupTheory/Congruence): simp lemmas for comap equivalences

### PR DESCRIPTION
**Pull request by hannahfechtner**
add a few lemmas, mostly about when the underlying types are equal, so too must be the quotient

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by tb65536** [comment]
Something's a bit messed up with this PR (the diff is 90k lines). Maybe merge master? Or if that doesn't work, ask on Zulip.

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (hannahfechtner) [reply Mathlib/GroupTheory/Congruence/Basic.lean]: you're right, it's fixed now! Looks like this material had been moved to Congrence/Defs while I was working on this PR, and I didn't fully update everything

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:78]: ~~I am surprised you need bijectivity here. I would expect this be true when `f` is not ...~~ Nevermind I am being silly

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:79]: Can you prove an inequality one way using only injectivity/surjectivity?

Comment (hannahfechtner) [reply Mathlib/GroupTheory/Congruence/Basic.lean:91]: Neat idea! I can prove one inequality without either, and then the other inequality requires bijectivity. I've separated out the inequality which doesn't require either as conGen_le_comap; but when I use it for the second half of comap_conGen_of_Bijective, it only unifies when I use "apply", not "exact" --
"show term" produces an "exact" statement that won't work. Do you have any idea what's happening?

Comment (hannahfechtner) [reply Mathlib/GroupTheory/Congruence/Basic.lean]: Thanks! I think I might've figured it out.

I think it's because of the implicit arguments for the instance of LE on Con
∀ ⦃x y⦄, c x y → d x y

(which here should be a and b)

Lean won't match them correctly with "exact"; it searches harder with "apply"

If I simply give them to "exact", by using @ to specify the implicit arguments as well, that works too

And even better, if I use a fact about LE to avoid using ext, everything clicks into place just right

Thanks for pointing out that the two directions should be separate - it makes the code much nicer!

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:82]: Here are a few uses of dot notation that might or might not work
```suggestion
  exact ConGen.Rel.rec (fun x y h ↦ ConGen.Rel.of (f x) (f y) h) (fun x ↦ .refl (f x))
```

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:83]:
```suggestion
    (fun _ ↦ .symm) (fun _ _ ↦ .trans) (fun {w x y z} _ _ h1 h2 ↦
```

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:86]:
```suggestion
    (h1.mul h2))) h
```

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:247]:
```suggestion
noncomputable def comapQuotientEquivOfSurj (c : Con M) (f : N →* M) (hf : Function.Surjective f) :
```

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:244]:
```suggestion
/-- If `e : M →* N` is surjective then `(c.comap e).Quotient ≃* c.Quotient` with `c : Con N`. -/
```
Same below

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:72]:
```suggestion
@[to_additive (attr := simp)]
```
same below

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:262]:
```suggestion
/-- This version infers the surjectivity of the function from a MulEquiv function -/
@[to_additive (attr := simp) "Copy edocstring here"]
lemma comapQuotientEquivOfSurj_symm_mk' (c : Con M) (f : N ≃* M) (x : N) :
```

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:264]:
```suggestion
      (comapQuotientEquivOfSurj c f f.surjective)) ⟦f x⟧) = ↑x :=
```

**Review by YaelDillies — COMMENTED** [review]
I have strain injuries to both wrists, hence must not type much. Please apologise the poor spelling, vague indications and general conciseness.

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (hannahfechtner) [reply Mathlib/GroupTheory/Congruence/Basic.lean]: done!

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (hannahfechtner) [reply Mathlib/GroupTheory/Congruence/Basic.lean]: added!

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (hannahfechtner) [reply Mathlib/GroupTheory/Congruence/Basic.lean]: copied!

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (hannahfechtner) [reply Mathlib/GroupTheory/Congruence/Basic.lean:265]: Happy to make the change! Could you explain what's happening here so I know for the future? 

I'd thought hypotheses beyond the first line were indented twice (2 tabs, or 4 spaces), and then the proof itself begins at one indent (1 tab, or 2 spaces). 

Would this mean 
(A) first line is not indented, second line with hypotheses is indented twice, third line is indented 3 times, fourth 4 times, etc?

(B) first line is not indented, all further lines with hypotheses are indented twice, but this one gets an extra indent to show the grouping of the parentheses? (basically since we're splitting up the parenthesized ```
(@MulEquiv.symm (Con.Quotient (comap ⇑f _ c)) _ _ _
    (comapQuotientEquivOfSurj c f f.surjective))
```)

trying to make it multiple choice to reduce typing for you :)

Comment (YaelDillies) [reply Mathlib/GroupTheory/Congruence/Basic.lean:265]: 😁 thanks for the thought. Basically A, except that later lines can be less indented than what A says depending on hestructure of the expression

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:249]:
```suggestion
  (Con.congr Con.comap_eq).trans <| Con.quotientKerEquivOfSurjective (c.mk'.comp f) <|
    Con.mk'_surjective.comp hf
```

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:257]: To show hf can be inferred
```suggestion
lemma comapQuotientEquivOfSurj_symm_mk (c : Con M) {f : N →* M} (hf) (x : N) :
    (comapQuotientEquivOfSurj c f hf).symm (f x) = x :=
```

Comment (hannahfechtner) [reply Mathlib/GroupTheory/Congruence/Basic.lean:257]: That's cool! Will the user still see the type of hf explicitly in mathlib? I guess my question is, this seems slick, but is it making the lemma harder to use? It may be because I'm a beginner, but I like to know at-a-glance what hypotheses a lemma or theorem requires when I'm on the leanprover-community site

Comment (YaelDillies) [reply Mathlib/GroupTheory/Congruence/Basic.lean:257]: No, it's producing exactly the same lemma with exactly the same hover etc

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:103]:
```suggestion
      rwa [fa, fb]
```

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:85]:
```suggestion
    (((congrArg (fun a ↦ conGen rel (f w * f y) a) (H x z))).mpr
```

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Congruence/Basic.lean:84]:
```suggestion
    (congrArg (fun a ↦ conGen rel a (f (x * z))) (H w y)).mpr
```

**Review by YaelDillies — APPROVED** [review]
maintainer merge

apart from that. Thanks!

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (Vierkantor) [inline Mathlib/GroupTheory/Congruence/Basic.lean:90]: I feel like the proof below can be neater if we first prove it for a `MulEquiv` and then suit the hypotheses here using `MulEquiv.ofBijective`. Let me see if I can cook up something...

Comment (Vierkantor) [reply Mathlib/GroupTheory/Congruence/Basic.lean:90]: Here's how I would use a `MulEquiv`. (Unfortunately I can't figure out how to make this suggestion...)

```lean
@[to_additive]
theorem comap_conGen_equiv {M N : Type*} [Mul M] [Mul N] (f : MulEquiv M N) (rel : N → N → Prop) :
    Con.comap f (map_mul f) (conGen rel) = conGen (fun x y ↦ rel (f x) (f y)) := by
  apply le_antisymm _ (le_comap_conGen f (map_mul f) rel)
  intro a b h
  simp only [Con.comap_rel] at h
  have H : ∀ n1 n2, (conGen rel) n1 n2 → ∀ a b, f a = n1 → f b = n2 →
      (conGen fun x y ↦ rel (f x) (f y)) a b := by
    intro n1 n2 h
    induction h with
    | of x y h =>
      intro _ _ fa fb
      apply ConGen.Rel.of
      rw [fa, fb]
      exact h
    | refl x =>
      intro _ _ fc fd
      rw [f.injective (fc.trans fd.symm)]
      exact ConGen.Rel.refl _
    | symm _ h => exact fun a b fs fb ↦ ConGen.Rel.symm (h b a fb fs)
    | trans _ _ ih ih1 =>
      exact fun a b fa fb ↦ Exists.casesOn (f.surjective _) fun c' hc' ↦
      ConGen.Rel.trans (ih a c' fa hc') (ih1 c' b hc' fb)
    | mul _ _ ih ih1 =>
      rename_i w x y z _ _
      intro a b fa fb
      rw [← f.eq_symm_apply, map_mul] at fa fb
      rw [fa, fb]
      exact ConGen.Rel.mul (ih (f.symm w) (f.symm x) (by simp) (by simp))
        (ih1 (f.symm y) (f.symm z) (by simp) (by simp))
  exact H (f a) (f b) h a b (refl _) (refl _)

@[to_additive]
theorem comap_conGen_of_bijective {M N : Type*} [Mul M] [Mul N] (f : M → N)
    (hf : Function.Bijective f) (H : ∀ (x y : M), f (x * y) = f x * f y) (rel : N → N → Prop) :
    Con.comap f H (conGen rel) = conGen (fun x y ↦ rel (f x) (f y)) :=
  comap_conGen_equiv (MulEquiv.ofBijective (MulHom.mk f H) hf) rel
```

The alternative would be to put the creation of a `MulEquiv` inline in the `mul` case of the induction; I think this is neater since it means less argument reshuffling for users who already have a `MulEquiv`.

**Review by Vierkantor — APPROVED** [review]
I'm happy with this but since I made a big suggestion, I'll ask another maintainer to take a final look.

**Comment by Vierkantor** [comment]
Discussed with @jcommelin and approved so:

bors d+

File: Mathlib/GroupTheory/Congruence/Basic.lean
Comment (hannahfechtner) [reply Mathlib/GroupTheory/Congruence/Basic.lean:90]: oh nice, thanks!

**Comment by Vierkantor** [comment]
Oh, it looks like my change wasn't yet applied. So we need to wait before merging until it is done. Is it okay if I push to your branch?

bors r-

**Comment by hannahfechtner** [comment]
@Vierkantor of course! thank you

**Comment by Vierkantor** [comment]
Pushed and built so Let's Get This Merged!

bors r+

**Pull request closed without merging**
